Introducing Swift C++ Interop in WebKit

nmahendru opened the pull request:
#### 29b3336cfd7aa527604d5e5087331bfc7d4306e5
<pre>
Introducing Swift C++ Interop in WebKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=271263">https://bugs.webkit.org/show_bug.cgi?id=271263</a>
<a href="https://rdar.apple.com/123331949">rdar://123331949</a>

Reviewed by Elliott Williams.

This fixes the build failures introduced by 276651@main.
(That commit was reverted in 276659@main.)

The CryptoKit Changes are still hidden behind a feature flag
`CryptoKitEnabled`.

This is an examplar of the C++ Swift interop feature.
We have randomly chosen the Digest Implementation to be the first one to adopt this.
The Compiler feature is only available for newer platforms to code has been guarded
with ifdefs.

The changes in AbortWithReasonSPI.h are to allow the Swift Generator to work.
It, at the moment cannot handle headers included inside `extern C` blocks.

Input parameteres to Swift have been intentionally chosen as spans as
swift, at the moment my make copies of inputs for safety guarantees.
Copying a span is cheap so that is Okie.
On the return path, Swift calls the WTF::Vector move constructor(confirmed in debugger).

* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/spi/darwin/AbortWithReasonSPI.h:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
(AesGcm.test(_:)):
(AesKwRV.errCode):
(AesKwRV.outputSize):
(HashFunction.update(_:)):
(Digest.sha1(_:)):
(Digest.sha256(_:)):
(Digest.sha384(_:)):
(Digest.sha512(_:)):
(AesKwReturnValue.errCode): Deleted.
(AesKwReturnValue.outputSize): Deleted.
* Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift: Added.
(HashFunction.update(_:)):
* Source/WebCore/PAL/pal/PALSwiftModule.h: Copied from Source/WTF/wtf/spi/darwin/AbortWithReasonSPI.h.
* Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.cpp: Renamed from Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.mm.
(PAL::CryptoDigest::computeHash):
* Source/WebCore/PAL/pal/crypto/openssl/CryptoDigestOpenSSL.cpp:
(PAL::createCryptoDigest):
(PAL::CryptoDigest::computeHash):
* Source/WebCore/PAL/pal/module.modulemap: Added.
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/crypto/CryptoAlgorithm.cpp:
(WebCore::CryptoAlgorithm::dispatchDigest):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.cpp: Renamed from Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.mm.
(WebCore::encryptCryptoKitAESGCM):
(WebCore::CryptoAlgorithmAESGCM::platformEncrypt):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.cpp: Renamed from Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.mm.
(WebCore::wrapKeyAESKWCryptoKit):
(WebCore::unwrapKeyAESKWCryptoKit):
* Source/WebCore/crypto/cocoa/CryptoUtilitiesCocoa.cpp:

Originally-landed-as: 276651@main (e88f01ab1a57). <a href="https://rdar.apple.com/123331949">rdar://123331949</a>
Canonical link: <a href="https://commits.webkit.org/276767@main">https://commits.webkit.org/276767@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b9d2d9f96ea97a7b2c7c720de1c313ee4d01744b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45638 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24764 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48216 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48306 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41673 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47944 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29020 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22160 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/37376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46216 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21833 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/39364 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18521 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/19230 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40455 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3682 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/38866 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41934 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40785 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/50063 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/45106 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20626 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/17140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44468 "layout-tests (failure)") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/21932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43306 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10137 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/22292 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/52263 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21619 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10683 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

nmahendru commented on Source/WebCore/PAL/Configurations/PAL.xcconfig:
```diff
@@ -66,6 +76,33 @@ AVIF_LIBTOOLFLAGS_YES = $(BUILT_PRODUCTS_DIR)/libavif.a;
+OTHER_SWIFT_FLAGS_appletvsimulator = $(OTHER_SWIFT_FLAGS$(WK_TVOS_17));
+OTHER_SWIFT_FLAGS_watchos = $(OTHER_SWIFT_FLAGS$(WK_WATCHOS_10));
+OTHER_SWIFT_FLAGS_watchsimulator = $(OTHER_SWIFT_FLAGS$(WK_WATCHOS_10));
+OTHER_SWIFT_FLAGS_visionos = -Xcc -std=c++20 -I $(SRCROOT)/pal -I $(SRCROOT)/PAL/pal;
```
@emw-apple should this be `xros` ? or `visionos` ?

emw-apple replied:
xros. If you ever need to look this up, it's defined at the root of the SDK in `SDKSettings.json`.

nmahendru committed:
Introducing Swift C++ Interop in WebKit
https://bugs.webkit.org/show_bug.cgi?id=271263
rdar://123331949

Reviewed by Elliott Williams.

This fixes the build failures introduced by 276651@main.
(That commit was reverted in 276659@main.)

The CryptoKit Changes are still hidden behind a feature flag
`CryptoKitEnabled`.

This is an examplar of the C++ Swift interop feature.
We have randomly chosen the Digest Implementation to be the first one to adopt this.
The Compiler feature is only available for newer platforms to code has been guarded
with ifdefs.

The changes in AbortWithReasonSPI.h are to allow the Swift Generator to work.
It, at the moment cannot handle headers included inside `extern C` blocks.

Input parameteres to Swift have been intentionally chosen as spans as
swift, at the moment my make copies of inputs for safety guarantees.
Copying a span is cheap so that is Okie.
On the return path, Swift calls the WTF::Vector move constructor(confirmed in debugger).

* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/spi/darwin/AbortWithReasonSPI.h:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
(AesGcm.test(_:)):
(AesKwRV.errCode):
(AesKwRV.outputSize):
(HashFunction.update(_:)):
(Digest.sha1(_:)):
(Digest.sha256(_:)):
(Digest.sha384(_:)):
(Digest.sha512(_:)):
(AesKwReturnValue.errCode): Deleted.
(AesKwReturnValue.outputSize): Deleted.
* Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift: Added.
(HashFunction.update(_:)):
* Source/WebCore/PAL/pal/PALSwiftModule.h: Copied from Source/WTF/wtf/spi/darwin/AbortWithReasonSPI.h.
* Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.cpp: Renamed from Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.mm.
(PAL::CryptoDigest::computeHash):
* Source/WebCore/PAL/pal/crypto/openssl/CryptoDigestOpenSSL.cpp:
(PAL::createCryptoDigest):
(PAL::CryptoDigest::computeHash):
* Source/WebCore/PAL/pal/module.modulemap: Added.
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/crypto/CryptoAlgorithm.cpp:
(WebCore::CryptoAlgorithm::dispatchDigest):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.cpp: Renamed from Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.mm.
(WebCore::encryptCryptoKitAESGCM):
(WebCore::CryptoAlgorithmAESGCM::platformEncrypt):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.cpp: Renamed from Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.mm.
(WebCore::wrapKeyAESKWCryptoKit):
(WebCore::unwrapKeyAESKWCryptoKit):
* Source/WebCore/crypto/cocoa/CryptoUtilitiesCocoa.cpp:

Originally-landed-as: 276651@main (e88f01ab1a57). rdar://123331949
Canonical link: https://commits.webkit.org/276767@main
